src/clients/java/src/client.zig

```diff
@@ -332,7 +346,6 @@ const NativeClient = struct {
         _ = client;
 
         var context = @intToPtr(*JNIContext, context_ptr);
-        defer context.release_packet(packet);
```

Comment:
Heh, this is subtle! Tbh, I am a bit unnerved at how we need to release this on the java side, but this indeed seems subtle: we want to release _after_ we've done memcpy, but _before_ we return from the Java method.

I am wondering, can we perhaps push `Request.memcpy` into zig code, so that, when we call `Request.endRequest`, it only gets JVM-owned objects?

Reply:
Ah, I think another way we can do that is, instead of introducing `release`, split `end_request` into `copy_result` and `end_request`. So that the sequence of calls on the native side is

```
callJvm(“copy_result”)
release(packet)
callJVM(“end_request”)
```

the main thing I want to avoid here is splitting resource management between native and JVM, as that seems error prone.

Reply:
That's an interesting bug.

We want to avoid switching between Java and JNI worlds as much as possible because [it interferes with JVM's ability to optimize code](https://developer.ibm.com/articles/j-jni/#choosing). With that in mind, differently from Go and C# clients, in Java we group a lot of things (e.g. acquiring a packet, populating and submitting it) into a single JNI call.

>I am wondering, can we perhaps push Request.memcpy into zig code, so that, when we call Request.endRequest, it only gets JVM-owned objects?

This change only exposes `tb_client_release_packet` (+1 JNI call), following the same logic as the other clients. However, it still keeps `tb_client_acquire_packet` hidden from the JVM, so it would make sense to expose something other than `release` as you suggested. We just have to make sure that we are not introducing too many JNI boundaries.

(...)

This is where we release the packet from the Java side in this PR (replacing `defer` at the zig side):
https://github.com/tigerbeetledb/tigerbeetle/blob/c1cc4b53b81a2341069467ade49617f2431c3f7c/src/clients/java/src/main/java/com/tigerbeetle/Request.java#L141-L145

And these are how we do exactly the same in C# and Go:
https://github.com/tigerbeetledb/tigerbeetle/blob/9703bc2e2a383e39a99bd1e15cb02dfb01282740/src/clients/dotnet/TigerBeetle/Request.cs#L117-L120

https://github.com/tigerbeetledb/tigerbeetle/blob/9703bc2e2a383e39a99bd1e15cb02dfb01282740/src/clients/go/tb_client.go#L178-L179

It's worth noting that this bug always existed, but it was silent before #793  because we didn't have `ConcurrenctExceededException`, instead, we just blocked until a packet became available.

Reply:
>And these are how we do exactly the same in C# and Go:

Ok, I think if we do explicit release everywhere, it's probably better that way!

Reply:
I liked the idea of doing memcpy on the JNI side and then releasing the package.
It would solve both the issues we are talking about in this PR.